# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: uriloader/exthandler/ExternalHelperAppParent.cpp
# Commit: baffa8a7b961
# Full Hash: baffa8a7b9618e9a02e4bd808adb71dead99cce5
# Author: Matt Woodrow <mwoodrow@mozilla.com>
# Date: 2019-11-12 09:43:07
# Regressor Bug: 1589270
# File Overlap Count: 1
# Description:
#   Bug 1589270 - Part 3: Convert nsExternalHelperApp to use BrowsingContext instead of nsIInterfaceRequestor. r=bzbarsky
#   
#   This also converts MaybeCloseWindowHelper, and results in the window close operations being always run in the parent (even without DocumentChannel).
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D49528
# ==============================================================================

diff -r 68bc2ae9b89a -r baffa8a7b961 uriloader/exthandler/ExternalHelperAppParent.cpp
--- a/uriloader/exthandler/ExternalHelperAppParent.cpp	Fri Nov 08 03:01:03 2019 +0000
+++ b/uriloader/exthandler/ExternalHelperAppParent.cpp	Fri Nov 08 04:35:05 2019 +0000
@@ -19,6 +19,8 @@
 #include "nsNetUtil.h"
 #include "mozilla/dom/Document.h"
 #include "mozilla/net/ChannelDiverterParent.h"
+#include "mozilla/dom/CanonicalBrowsingContext.h"
+#include "mozilla/dom/WindowGlobalParent.h"
 
 #include "mozilla/Unused.h"
 
@@ -61,33 +63,11 @@
   }
 }
 
-already_AddRefed<nsIInterfaceRequestor> GetWindowFromBrowserParent(
-    PBrowserParent* aBrowser) {
-  if (!aBrowser) {
-    return nullptr;
-  }
-
-  nsCOMPtr<nsIInterfaceRequestor> window;
-  BrowserParent* browserParent = BrowserParent::GetFrom(aBrowser);
-  if (browserParent->GetOwnerElement()) {
-    window = do_QueryInterface(
-        browserParent->GetOwnerElement()->OwnerDoc()->GetWindow());
-  }
-
-  return window.forget();
-}
-
-void UpdateContentContext(nsIStreamListener* aListener,
-                          PBrowserParent* aBrowser) {
-  MOZ_ASSERT(aListener);
-  nsCOMPtr<nsIInterfaceRequestor> window = GetWindowFromBrowserParent(aBrowser);
-  static_cast<nsExternalAppHandler*>(aListener)->SetContentContext(window);
-}
-
 void ExternalHelperAppParent::Init(
     const Maybe<mozilla::net::LoadInfoArgs>& aLoadInfoArgs,
     const nsCString& aMimeContentType, const bool& aForceSave,
-    const Maybe<URIParams>& aReferrer, PBrowserParent* aBrowser) {
+    const Maybe<URIParams>& aReferrer, BrowsingContext* aContext,
+    const bool& aShouldCloseWindow) {
   mozilla::ipc::LoadInfoArgsToLoadInfo(aLoadInfoArgs,
                                        getter_AddRefs(mLoadInfo));
 
@@ -100,21 +80,23 @@
     SetPropertyAsInterface(NS_LITERAL_STRING("docshell.internalReferrer"),
                            referrer);
 
-  nsCOMPtr<nsIInterfaceRequestor> window;
-  if (aBrowser) {
-    BrowserParent* browserParent = BrowserParent::GetFrom(aBrowser);
-    if (browserParent->GetOwnerElement())
-      window = do_QueryInterface(
-          browserParent->GetOwnerElement()->OwnerDoc()->GetWindow());
-
-    bool isPrivate = false;
-    nsCOMPtr<nsILoadContext> loadContext = browserParent->GetLoadContext();
-    loadContext->GetUsePrivateBrowsing(&isPrivate);
-    SetPrivate(isPrivate);
+  WindowGlobalParent* parent = aContext->Canonical()->GetCurrentWindowGlobal();
+  if (parent) {
+    RefPtr<BrowserParent> browser = parent->GetBrowserParent();
+    if (browser) {
+      bool isPrivate = false;
+      nsCOMPtr<nsILoadContext> loadContext = browser->GetLoadContext();
+      loadContext->GetUsePrivateBrowsing(&isPrivate);
+      SetPrivate(isPrivate);
+    }
   }
 
-  helperAppService->DoContent(aMimeContentType, this, window, aForceSave,
-                              nullptr, getter_AddRefs(mListener));
+  helperAppService->CreateListener(aMimeContentType, this, aContext, aForceSave,
+                                   nullptr, getter_AddRefs(mListener));
+
+  if (mListener && aShouldCloseWindow) {
+    mListener->SetShouldCloseWindow();
+  }
 }
 
 void ExternalHelperAppParent::ActorDestroy(ActorDestroyReason why) {
@@ -128,12 +110,10 @@
 }
 
 mozilla::ipc::IPCResult ExternalHelperAppParent::RecvOnStartRequest(
-    const nsCString& entityID, PBrowserParent* contentContext) {
+    const nsCString& entityID) {
   MOZ_ASSERT(!mDiverted,
              "child forwarding callbacks after request was diverted");
 
-  UpdateContentContext(mListener, contentContext);
-
   mEntityID = entityID;
   mPending = true;
   mStatus = mListener->OnStartRequest(this);
@@ -171,9 +151,8 @@
 }
 
 mozilla::ipc::IPCResult ExternalHelperAppParent::RecvDivertToParentUsing(
-    PChannelDiverterParent* diverter, PBrowserParent* contentContext) {
+    PChannelDiverterParent* diverter) {
   MOZ_ASSERT(diverter);
-  UpdateContentContext(mListener, contentContext);
   auto p = static_cast<mozilla::net::ChannelDiverterParent*>(diverter);
   p->DivertTo(this);
 #ifdef DEBUG